app/services/caching_service.py

```
@@ -0,0 +1,512 @@
+                or a hex-encoded pickle representation for other `data_type` values.
+            data_type (str): Format of `data`. Use "json" to parse JSON; any other value treats `data`
+                as a hex-encoded pickle.
+
+        Returns:
+            Any: The deserialized Python object, or `None` if deserialization fails.
+        """
+        try:
+            if data_type == "json":
+                return json.loads(data)
+            else:
+                return pickle.loads(bytes.fromhex(data))
```

Ruff (0.13.2)

`pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)